File: fm-server/src/main/java/io/zeitwert/fm/server/config/crnk/CrnkConfiguration.java
package io.zeitwert.fm.server.config.crnk;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import io.crnk.core.boot.CrnkBoot;
import io.crnk.core.boot.CrnkProperties;
import io.crnk.core.module.SimpleModule;
import io.crnk.spring.internal.SpringServiceDiscovery;
import jakarta.servlet.*;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.io.IOException;
import java.util.List;

/**
 * Manual Crnk configuration for Spring Boot 3.
 * <p>
 * Since crnk-setup-spring-boot doesn't support Spring Boot 3, we configure
 * CrnkBoot manually using crnk-core and crnk-setup-spring modules.
 * <p>
 * Note: crnk-servlet uses javax.servlet which is incompatible with Spring Boot 3's
 * jakarta.servlet, so we create a custom Jakarta-compatible filter.
 */
@Configuration
public class CrnkConfiguration {

	@Value("${crnk.path-prefix:/api}")
	private String pathPrefix;

	@Value("${crnk.default-page-limit:50}")
	private Long defaultPageLimit;

	@Value("${crnk.max-page-limit:1000}")
	private Long maxPageLimit;

	@Value("${crnk.allow-unknown-attributes:false}")
	private boolean allowUnknownAttributes;

	@Value("${crnk.return404-on-null:true}")
	private boolean return404OnNull;

	@Bean
	public CrnkBoot crnkBoot(ApplicationContext applicationContext,
													 DefaultQuerySpecUrlMapper urlMapper,
													 List<CrnkBootConfigurer> configurers) {

		CrnkBoot boot = new CrnkBoot();

		// Set up Spring service discovery for finding Crnk repositories and modules
		SpringServiceDiscovery serviceDiscovery = new SpringServiceDiscovery();
		serviceDiscovery.setApplicationContext(applicationContext);
		boot.setServiceDiscovery(serviceDiscovery);

		// Configure web path prefix
		boot.setWebPathPrefix(pathPrefix);

		// Configure URL mapper (with custom path resolver)
		boot.setUrlMapper(urlMapper);

		// Configure properties
		boot.setPropertiesProvider(key -> {
			return switch (key) {
				case CrnkProperties.WEB_PATH_PREFIX -> pathPrefix;
				case CrnkProperties.ALLOW_UNKNOWN_ATTRIBUTES -> String.valueOf(allowUnknownAttributes);
				case CrnkProperties.RETURN_404_ON_NULL -> String.valueOf(return404OnNull);
				default -> null;
			};
		});

		// Configure pagination
		boot.setDefaultPageLimit(defaultPageLimit);
		boot.setMaxPageLimit(maxPageLimit);

		// Register a simple module for any additional setup
		SimpleModule simpleModule = new SimpleModule("zeitwert");
		boot.addModule(simpleModule);

		// Allow configurers to customize CrnkBoot
		configurers.forEach(configurer -> configurer.configure(boot));

		// Configure ObjectMapper for Spring Boot 3.x / Jackson 2.19.x compatibility
		ObjectMapper objectMapper = boot.getObjectMapper();
		// Ignore unknown properties (newer Jackson is stricter about unknown fields)
		objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
		// Register Java 8 date/time module for OffsetDateTime, LocalDate, etc.
		objectMapper.registerModule(new JavaTimeModule());
		// Write dates as ISO strings, not timestamps
		objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);

		// Initialize Crnk
		boot.boot();

		return boot;
	}

	@Bean
	public Filter crnkFilter(CrnkBoot crnkBoot) {
		// Create a Jakarta-compatible filter that wraps crnk's functionality
		return new JakartaCrnkFilter(crnkBoot, pathPrefix);
	}

	/**
	 * Jakarta Servlet API compatible filter for Crnk.
	 * <p>
	 * Since the original CrnkFilter uses javax.servlet (incompatible with Spring Boot 3),
	 * this filter provides the same functionality using jakarta.servlet.
	 */
		private record JakartaCrnkFilter(CrnkBoot crnkBoot, String pathPrefix) implements Filter {

		@Override
			public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
					throws IOException, ServletException {

				if (request instanceof HttpServletRequest httpRequest
						&& response instanceof HttpServletResponse httpResponse) {

					String path = httpRequest.getRequestURI();
					String contextPath = httpRequest.getContextPath();
					if (contextPath != null && !contextPath.isEmpty()) {
						path = path.substring(contextPath.length());
					}

					// Only process requests to the crnk API path
					if (path.startsWith(pathPrefix)) {
						// Create and process the request through crnk
						JakartaHttpRequestContext sessionContext = new JakartaHttpRequestContext(
								httpRequest, httpResponse, pathPrefix);

						// Process the request using crnk's request dispatcher
						crnkBoot.getRequestDispatcher().process(sessionContext);

						// Check if crnk handled the request
						if (sessionContext.hasResponse()) {
							sessionContext.flushResponse();
							return; // Crnk handled the request
						}
					}
				}

				// Pass to next filter if not handled by crnk
				chain.doFilter(request, response);
			}

		}

}

File: fm-server/src/main/java/io/zeitwert/fm/server/config/crnk/CrnkConfiguration.kt
package io.zeitwert.fm.server.config.crnk

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import io.crnk.core.boot.CrnkBoot
import io.crnk.core.boot.CrnkProperties
import io.crnk.core.engine.properties.PropertiesProvider
import io.crnk.core.module.SimpleModule
import io.crnk.spring.internal.SpringServiceDiscovery
import jakarta.servlet.Filter
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.io.IOException
import java.util.function.Consumer

/**
 * Manual Crnk configuration for Spring Boot 3.
 *
 * Note: crnk-servlet uses javax.servlet which is incompatible with Spring Boot 3's
 * jakarta.servlet, so we create a custom Jakarta-compatible filter.
 */
@Configuration
open class CrnkConfiguration {

	@Value("\${crnk.path-prefix:/api}")
	private val pathPrefix: String? = null

	@Value("\${crnk.default-page-limit:50}")
	private val defaultPageLimit: Long? = null

	@Value("\${crnk.max-page-limit:1000}")
	private val maxPageLimit: Long? = null

	@Value("\${crnk.allow-unknown-attributes:false}")
	private val allowUnknownAttributes = false

	@Value("\${crnk.return404-on-null:true}")
	private val return404OnNull = false

	@Bean
	open fun crnkBoot(
		applicationContext: ApplicationContext,
		urlMapper: DefaultQuerySpecUrlMapper?,
		configurers: MutableList<CrnkBootConfigurer>,
	): CrnkBoot {
		val boot = CrnkBoot()

		// Set up Spring service discovery for finding Crnk repositories and modules
		val serviceDiscovery = SpringServiceDiscovery()
		serviceDiscovery.setApplicationContext(applicationContext)
		boot.setServiceDiscovery(serviceDiscovery)

		// Configure web path prefix
		boot.webPathPrefix = pathPrefix

		// Configure URL mapper (with custom path resolver)
		boot.setUrlMapper(urlMapper)

		// Configure properties
		boot.setPropertiesProvider(
			PropertiesProvider { key: String? ->
				when (key) {
					CrnkProperties.WEB_PATH_PREFIX -> pathPrefix
					CrnkProperties.ALLOW_UNKNOWN_ATTRIBUTES -> allowUnknownAttributes.toString()
					CrnkProperties.RETURN_404_ON_NULL -> return404OnNull.toString()
					else -> null
				}
			},
		)

		// Configure pagination
		boot.setDefaultPageLimit(defaultPageLimit)
		boot.setMaxPageLimit(maxPageLimit)

		// Register a simple module for any additional setup
		val simpleModule = SimpleModule("zeitwert")
		boot.addModule(simpleModule)

		// Allow configurers to customize CrnkBoot
		configurers.forEach(Consumer { configurer: CrnkBootConfigurer? -> configurer!!.configure(boot) })

		// Configure ObjectMapper for Spring Boot 3.x / Jackson 2.19.x compatibility
		val objectMapper = boot.getObjectMapper()
		// Ignore unknown properties (newer Jackson is stricter about unknown fields)
		objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
		// Register Java 8 date/time module for OffsetDateTime, LocalDate, etc.
		objectMapper.registerModule(JavaTimeModule())
		// Write dates as ISO strings, not timestamps
		objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)

		// Initialize Crnk
		boot.boot()

		return boot
	}

	@Bean
	open fun crnkFilter(crnkBoot: CrnkBoot): Filter {
		// Create a Jakarta-compatible filter that wraps crnk's functionality
		return JakartaCrnkFilter(crnkBoot, pathPrefix)
	}

	/**
	 * Jakarta Servlet API compatible filter for Crnk.
	 *
	 * Since the original CrnkFilter uses javax.servlet (incompatible with Spring Boot 3),
	 * this filter provides the same functionality using jakarta.servlet.
	 */
	@JvmRecord
	private data class JakartaCrnkFilter(
		val crnkBoot: CrnkBoot?,
		val pathPrefix: String?,
	) : Filter {

		@Throws(IOException::class, ServletException::class)
		override fun doFilter(
			request: ServletRequest?,
			response: ServletResponse?,
			chain: FilterChain,
		) {
			if (request is HttpServletRequest &&
				response is HttpServletResponse
			) {
				var path = request.requestURI
				val contextPath = request.contextPath
				if (contextPath != null && !contextPath.isEmpty()) {
					path = path.substring(contextPath.length)
				}

				// Only process requests to the crnk API path
				if (path.startsWith(pathPrefix!!)) {
					// Create and process the request through crnk
					val sessionContext = JakartaHttpRequestContext(request, response, pathPrefix)
					// Process the request using crnk's request dispatcher
					try {
						crnkBoot!!.getRequestDispatcher().process(sessionContext)
					} catch (ex: Exception) {
						throw ServletException("Error processing Crnk request", ex)
					}
					// Check if crnk handled the request
					if (sessionContext.hasResponse()) {
						sessionContext.flushResponse()
						return // Crnk handled the request
					}
				}
			}

			// Pass to next filter if not handled by crnk
			chain.doFilter(request, response)
		}
	}

}
